File: examples/coagents/research-canvas/agent-js/src/delete.ts
/**
 * Delete Resources
 */

import { AgentState } from "./state";
import { RunnableConfig } from "@langchain/core/runnables";
import { ToolMessage, AIMessage } from "@langchain/core/messages";

export async function delete_node(
  state: AgentState,
  config: RunnableConfig
): Promise<AgentState> {
  /**
   * Delete Node
   */
  return state;
}

export async function perform_delete_node(
  state: AgentState,
  config: RunnableConfig
) {
  /**
   * Perform Delete Node
   */
  const aiMessage = state["messages"][
    state["messages"].length - 2
  ] as AIMessage;
  const toolMessage = state["messages"][
    state["messages"].length - 1
  ] as ToolMessage;

  let resources = state["resources"];

  if (toolMessage.content === "YES") {
    let urls: string[];

    if (aiMessage.tool_calls) {
      urls = aiMessage.tool_calls[0].args.urls;
    } else {
      const parsedToolCall = JSON.parse(
        aiMessage.additional_kwargs!.function_call!.arguments
      );
      urls = parsedToolCall.urls;
    }

    resources = resources.filter((resource) => !urls.includes(resource.url));
  }

  return {
    resources,
  };
}

File: examples/coagents/research-canvas/agent-py/research_canvas/crewai/delete.py
"""
Delete Resources
"""

import json
from typing_extensions import Dict, Any

def maybe_perform_delete(state: Dict[str, Any]):
    """
    Maybe perform delete.
    """
    messages = state["messages"]
    if len(messages) >= 2:
        last_message = messages[-1]
        prev_message = messages[-2]
        if (prev_message.get("tool_calls") and
            prev_message["tool_calls"][0]["function"].get("name") == "DeleteResources" and
            last_message.get("content") == "YES"):
            urls = json.loads(prev_message["tool_calls"][0]["function"]["arguments"])["urls"]
            state["resources"] = [
                resource for resource in state["resources"] if resource["url"] not in urls
            ]
